one-table.redshift.sql

```diff
@@ -0,0 +1,330 @@
+    , "id" int8 -- PK cannot be null, but after raw insert and before typing, row will be null
+    , "first_name" text
+    , "age" int8
+    , "address" super
+    , "updated_at" timestamp
+);
+
+---------------------------------------
+--------- SAFE CAST METHODS -----------
+---------------------------------------
+
+-- DROP FUNCTION _airbyte_safe_cast_to_integer(input text)
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_integer(input text)
+RETURNS INTEGER
+STABLE AS $$
+	try:
+		v = int(input)
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
+
+-- DROP FUNCTION _airbyte_safe_cast_to_boolean(input text)
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_boolean(input text)
+RETURNS BOOLEAN
+STABLE AS $$
+	try:
+		v = input.lower() == 'true'
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
+
+-- DROP FUNCTION _airbyte_safe_cast_to_text(input text);
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_text(input text)
+RETURNS TEXT
+STABLE AS $$
+	try:
+		v = str(input)
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
+
+-- DROP FUNCTION _airbyte_safe_cast_to_timestamp(input text);
+CREATE OR REPLACE FUNCTION _airbyte_safe_cast_to_timestamp(input text)
+RETURNS TIMESTAMP
+STABLE AS $$
+	import dateutil.parser
+	try:
+		v = dateutil.parser.parse(input)
+		return v
+	except:
+		return None
+$$ LANGUAGE plpythonu;
+
+---------------------------
+--------- UTILS -----------
+---------------------------
+
+-- http://datasamurai.blogspot.com/2016/06/creating-uuid-function-in-redshift.html
+CREATE OR REPLACE FUNCTION gen_uuid()
+RETURNS VARCHAR(36)
+VOLATILE AS $$
+ import uuid
+ return uuid.uuid4().__str__()
+$$ LANGUAGE plpythonu;
+
+-------------------------------------
+--------- TYPE AND DEDUPE -----------
+-------------------------------------
+
+CREATE OR REPLACE PROCEDURE _airbyte_prepare_raw_table()
+AS $$
+BEGIN
+  CREATE SCHEMA IF NOT EXISTS x;
+  CREATE TABLE IF NOT EXISTS x.users_raw (
+      "_airbyte_raw_id" VARCHAR(36) NOT NULL, -- Airbyte column, cannot be null
+      "_airbyte_data" SUPER NOT NULL, -- Airbyte column, cannot be null
+      "_airbyte_extracted_at" timestamp NOT NULL, -- Airbyte column, cannot be null
+      "_airbyte_loaded_at" timestamp, -- Airbyte column
+      PRIMARY KEY ("_airbyte_raw_id")
+  );
+END
+$$ LANGUAGE plpgsql;
+
+
+CREATE OR REPLACE PROCEDURE _airbyte_type_dedupe()
+AS $$
+
+DECLARE missing_pk_count INTEGER DEFAULT 0;
+
+BEGIN
+
+  -- Step 1: Validate the incoming data
+  -- We can't really do this properly in the pure-SQL example here, but we should throw if any row doesn't have a PK
+  missing_pk_count := (
+    SELECT COUNT(1)
+    FROM x.USERS_RAW
+    WHERE
+      "_airbyte_loaded_at" IS NULL
+      AND _airbyte_safe_cast_to_integer(_airbyte_data.id::text) IS NULL
+    );
+
+  IF missing_pk_count > 0 THEN
+    RAISE EXCEPTION 'Table has % rows missing a primary key', missing_pk_count;
```

Comment:
does it mean the whole load will fail if one recodrd doesn't have a PK?

Reply:
Yep (and a missing cursor will fail syncs as well)!
If you don't have a PK or cursor, we can't reliably tell you anything about the sync - which row had a problem, or which row should win a dedupe.

Reply:
Any other column can be null